### PR TITLE
feat(table): drop dangling equality deletes during RewriteDataFiles

### LINKED ISSUE #946: feat(table): drop dangling equality delete files during RewriteDataFiles
### Feature Request / Improvement

When `Transaction.RewriteDataFiles` compacts data files, position delete files are removed from the new snapshot if they pointed at rewritten files (good!). 

Equality delete files are intentionally preserved (`collectSafePositionDeletes`, [`rewrite_data_files.go:201-235`](https://github.com/apache/iceberg-go/blob/main/table/rewrite_data_files.go#L201-L235)).

For high-churn CDC workloads this leaves dangling equality delete entries in every subsequent snapshot's manifests forever. Compaction restores read perf in the short term (the seq-number filter at [`scanner.go:372-398`](https://github.com/apache/iceberg-go/blob/main/table/scanner.go#L372-L398) skips them), but the manifest read overhead grows linearly with the number of preserved eq-delete entries. Over days of replication this becomes the dominant scan cost.

I'd like to propose extending the cleanup to also drop equality deletes that are provably dead after compaction.

### Why this matters

Workload: PostgreSQL → Iceberg v2 CDC replication, ~one snapshot every 5 seconds, ~1 equality delete file per snapshot for UPDATE/DELETE operations.

Empirical numbers from a small repro on a 10K-row table (Apple M1 Pro, local REST + MinIO):

| State | Files (data / eq-delete) | `COUNT(*)` scan |
|---|---|---|
| Clean | 1 / 0 | ~30ms (baseline) |
| 10K DML, no compaction | 16 / 120 | ~217ms (peak 382ms) |
| **After `iceberg compact run`** | **1 / 120** | **~50ms** |

So compaction is effective — 16 → 1 data file, scan time drops from 217ms to 50ms. But the 120 dangling eq-delete files now sit in every future snapshot's manifests with no purpose.

Extrapolating sustained CDC at one eq-delete per snapshot:

| Time running | Dangling eq-deletes | Estimated overhead |
|---|---|---|
| 1 hour | ~720 | ~120ms per scan |
| 1 day | ~17,000 | ~3 sec per scan |
| 1 week | ~120,000 | ~20 sec per scan |

`ExpireSnapshots` doesn't help here, because the **current** snapshot's manifests still reference the dead eq-delete files. The cleanup has to happen during the rewrite commit itself.

### When is an equality delete safe to drop?

The Iceberg v2 spec says an equality delete file E applies to a data file D iff `E.seq > D.seq` (and partitions match). After `RewriteDataFiles` rewrites a set of files in a partition, the new files have a sequence number strictly greater than every existing file in that partition.

So an equality delete file E is **provably dead** after compaction when:

1. Every data file with `data.seq < E.seq` in E's partition was in the rewrite set, AND
2. `E.seq < new_compacted_file.seq` (always true — compaction commit gets the new highest seq)

In other words: if compaction processed every data file E could have applied to, E will never apply to anything again.

The simplest case — and the one most CDC sinks hit — is **unpartitioned full-partition compaction**: when every data file in the (only) partition is being rewritten, every equality delete file in that partition becomes dead.

### PR DESCRIPTION
**Pull request by laskoviymishka**
Closes #946.

Sustained CDC workloads (Postgres → Iceberg replication, ~one snapshot per few seconds with one eq-delete per UPDATE/DELETE) accumulate equality-delete files in every snapshot's manifests forever. After compaction, every preexisting eq-delete is provably dead — no surviving applicable data file has seq < eq-delete.seq — but the old RewriteDataFiles only removed position deletes, leaving eq-deletes in place. Manifest fanout grew linearly with the CDC duration.

This change adds three pieces:

1. table/compaction.DecideDeadEqualityDeletes — a pure predicate that exactly matches the v2 reader (table/scanner.go matchEqualityDeletesToData): E applies to D iff E.seq > D.seq AND (either side has empty partition OR partition tuples match). SpecID is intentionally NOT in the predicate — using it would silently drop eq-deletes the reader still applies under partition-spec evolution.

2. table/compaction.CollectDeadEqualityDeletes — walks the snapshot's manifests via the existing public Snapshot.Manifests + FetchEntries APIs, builds a SurvivorSurvey, and returns the dead set. Two-pass design: skip the data-manifest walk if there are no eq-delete candidates.

3. table.RewriteDataFilesOptions.ExtraDeleteFilesToRemove — a new field on the options struct that the executor passes through to ReplaceFiles. The caller (cmd/iceberg/compact.go or any library user) computes dead eq-deletes via CollectDeadEqualityDeletes against the same snapshot and threads them in. The executor itself is policy-free.

The CLI gains a --preserve-dead-equality-deletes flag (default off, i.e., cleanup is on by default — recommended for sustained CDC).

The RewriteDataFiles signature collapses (partialProgress, snapshotProps, opts...) into a single RewriteDataFilesOptions struct. RewriteResult splits RemovedDeleteFiles into RemovedPositionDeleteFiles + RemovedEqualityDeleteFiles for observability.

Tests:
- table/compaction/eq_delete_decision_test.go pins the predicate against the scanner's rule across mixed-partition / cross-spec / boundary cases.
- table/compaction/cdc_stress_test.go runs 80 CDC cycles (81 data files + 80 eq-deletes pre-compaction → 1 data file + 0 eq-deletes post). Skipped under -short.
- table/rewrite_data_files_test.go covers full-rewrite drops, partial-rewrite preserves, partial-progress preserves (per-group cleanup is a follow-up), and the CLI opt-out path.

### CODE REVIEW

File: table/compaction/eq_delete_collect.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+	}
+
+	manifests, err := snap.Manifests(fs)
+	if err != nil {
+		return nil, err
+	}
+
+	// First pass: delete manifests only, collecting eq-delete candidates.
+	var candidates []iceberg.ManifestEntry
+	for _, m := range manifests {
```
Comment (zeroshade) [inline table/compaction/eq_delete_collect.go:67]: don't we have a helper for this to simplify the loops?

File: table/compaction/eq_delete_collect.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+			candidates = append(candidates, e)
+		}
+	}
+	if len(candidates) == 0 {
+		return nil, nil
+	}
+
+	// Second pass: data manifests, building the survivor survey.
+	survey := NewSurvivorSurvey()
+	for _, m := range manifests {
```
Comment (zeroshade) [inline table/compaction/eq_delete_collect.go:91]: same comment as above

File: table/compaction/eq_delete_decision.go [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+// consistently — typical pattern is `survey := NewSurvivorSurvey()`
+// followed by `survey.AddSurvivor(...)` which Go auto-addresses.
+func (s *SurvivorSurvey) AddSurvivor(partition map[int]any, seq int64) {
+	if seq < 0 {
+		seq = 0
+	}
+	if len(partition) == 0 {
+		if seq < s.EmptyPartMinSeq {
+			s.EmptyPartMinSeq = seq
+		}
```
Comment (zeroshade) [inline table/compaction/eq_delete_decision.go]:
```suggestion
		s.EmptyPartMinSeq = min(seq, s.EmptyPartMinSeq)
```

File: table/compaction/eq_delete_decision.go [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+// E (so EmptyPartMinSeq is always in the min); an unpartitioned E
+// applies to every surviving D (so we min across every PartMinSeq).
+func (s *SurvivorSurvey) applicableMinSeq(eqPartition map[int]any) int64 {
+	if len(eqPartition) == 0 {
+		min := s.EmptyPartMinSeq
+		for _, v := range s.PartMinSeq {
+			if v < min {
+				min = v
+			}
+		}
```
Comment (zeroshade) [inline table/compaction/eq_delete_decision.go]: `return min(s.EmptyPartMinSeq, slices.Min(maps.Values(s.PartMinSeq)))` ?

File: table/compaction/eq_delete_decision.go [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+//
+// Dedup by file path: the same eq-delete file may appear in multiple
+// manifest entries after manifest merging.
+func DecideDeadEqualityDeletes(survey *SurvivorSurvey, candidates []iceberg.ManifestEntry) []iceberg.DataFile {
+	if survey == nil {
+		return nil
+	}
+	if len(candidates) == 0 {
+		return nil
+	}
```
Comment (zeroshade) [inline table/compaction/eq_delete_decision.go]: just combine these into a single check `if survey == nil || len(candidates) == 0 { return nil }`

File: table/rewrite_data_files.go [hunk 6]
```diff
@@ -63,24 +71,54 @@ type CompactionTaskGroup struct {
...
-// requires committing separate transactions per group, which is left
-// to the caller. When partialProgress is false, all groups are committed
-// in a single atomic snapshot.
+// Equality-delete cleanup is the caller's responsibility: compute the
+// dead set with table/compaction.CollectDeadEqualityDeletes (against
+// the same snapshot the rewrite is staged on) and pass it via
+// opts.ExtraDeleteFilesToRemove. The executor only orchestrates the
+// commit; it does not impose a cleanup policy. This split keeps the
+// pure spec predicate in table/compaction and the unexported snapshot
+// machinery in table.
```
Comment (zeroshade) [inline table/rewrite_data_files.go]: surround the references with brackets (i.e. `[compaction.CollectDeadEqualityDeletes]`) so that the docstrings will cross link in pkg.go.dev

**Review by zeroshade — CHANGES_REQUESTED** [review]

File: table/compaction/eq_delete_collect.go (hunk 1 above)
Comment (laskoviymishka) [reply table/compaction/eq_delete_collect.go:67]: Switched FetchEntries(fs, false) → (fs, true);

File: table/compaction/eq_delete_collect.go (hunk 2 above)
Comment (laskoviymishka) [reply table/compaction/eq_delete_collect.go:91]: same

File: table/compaction/eq_delete_decision.go (hunk 4 above)
Comment (laskoviymishka) [reply table/compaction/eq_delete_decision.go]: Done, added slices.Collect because slices.Min takes a slice, not iter.Seq

File: table/compaction/eq_delete_decision.go (hunk 5 above)
Comment (laskoviymishka) [reply table/compaction/eq_delete_decision.go]: make sense

File: table/rewrite_data_files.go (hunk 6 above)
Comment (laskoviymishka) [reply table/rewrite_data_files.go]: fixed everywhere

**Review by zeroshade — APPROVED** [review]
